### DISCUSSION TITLE
Use kind as k8s CI tool

### DISCUSSION THREAD

**Discussion by yxxhero** [helmfile/helmfile#12] [Ideas]
Use kind as k8s CI tool

**Comment by mumoshu** [comment]
Why do we need to do so?

**Reply by yxxhero** [reply]
“Kubernetes in Docker” aka Kind is changing the way one can setup k8s cluster locally . Using this tool, we can spin up multiple local k8s clusters in a matter of few minutes, all using docker containers. This is ideal for local development and testing of kubernetes based applications. It adds value for testing Kubernetes.

### Below are some of the advantages of using kind
Developer friendly, and can be used for testing k8s apps. It can also be easily replicated.
We can spin up multi-node k8s clusters with HA support.
Supports building kubernetes release builds from source. Imagine the advantage of testing an upcoming k8s release, locally.
It supports Linux, macOS, and windows.
Make docker images easily available to kubernetes clusters.
Good documentation and very stable

### Compare Kind with Minikube

Minikube has been around more many years, which basically spins up a VM, and acts as a single node K8s cluster. So you require a hypervisor like VirtualBox to be running. As a DevOps, while giving a demo using minikube has its own performance impact, which takes up a lot of local system resources and can slow down the system while presenting something.
When I tried with Kind, the main advantage is its fast boot-up time. Even the initial setup and installation were completed in a matter of 2–3 minutes. It relies on a nested docker image which has the capability to run systemd, kubernetes, containerd etc. The kubernetes running inside the docker image uses containerd as runtime.

**Reply by mumoshu** [reply]
Thanks for summarizing!

I think I might know how kind is generally considered better than minikube.
I still use it in many of my projects! I even have experience switched from minikube to kind before.

Buf if you could let me be clear, minikube's docker provider(`minikube start --driver docker`) seems also great these days.
And minikube seems to have more contributors and seems to receive support for the latest Kubernetes version earlier if I see recent releases on both projects.

Feature-wise, I especially like `eval $(minikube docker-env)`, which allows you to build arbitrary images **within** the dockerd used by minikube.

In kind, we have to docker-build on the host machine and load it into kind nodes via `kind load (docker-image|image-archive)`. For large container images this takes several to dozens of seconds per image depending on your build machine.

Why the ability to load images faster important?

If we're going to use kind or minikube to set up a testbed for E2E tests, each E2E test run should be reliable, so that it's more clear if it failed due to transient failures(sometimes due to external container image registry's partial down time) or permanent failures(we need to fix it in the implementation).

To make it reliable, it would be effective to replicate the whole test environment locally, not only k8s clusters(backed by either kind or minikube), but also chart registries and container registries. I already figured out how to set up chart registries locally while maintaining https://github.com/variantdev/chartify. For the container registry, letting us building all the required images using `eval $(minikube docker-env)` would mostly omit the necessity to rely on external registries in E2E test runs, hence more reliable.

If it look very long time to load all the images, it can be impractical, and that's why I'm not fond of `kind load`. I tried several docker builds using `eval $(minikube docker-env)` and it seems to work reliably and fast so far. If we don't find any downsides in minikube compared to kind, I'd bet minikube is a better option for helmfile's case.
